Add the @TestTransaction annotation

stuartwdouglas opened the pull request:
This allows you to run tests in a rollback only transaction.

Fixes #6463

stuartwdouglas committed:
Add the @TestTransaction annotation

This allows you to run tests in a rollback only transaction.

Fixes #6463

mkouba commented on test-framework/common/src/main/java/io/quarkus/test/TestTransaction.java:
```diff
@@ -0,0 +1,21 @@
+/**
+ * Indicates that this method should be run in a rollback only transaction.
+ *
+ * The allows the test method to modify the database as required, and then have
```
"The allows" is a typo probably?

mkouba replied:
Typo still present...

mkouba commented on extensions/narayana-jta/deployment/src/main/java/io/quarkus/narayana/jta/deployment/NarayanaJtaProcessor.java:
```diff
@@ -96,6 +114,40 @@ public void build(NarayanaJtaRecorder recorder,
     }
 
+    @BuildStep
+    void testTx(LaunchModeBuildItem lm, BuildProducer<GeneratedBeanBuildItem> generatedBeanBuildItemBuildProducer) {
```
Hm, so you generate an interceptor class because you can't add the `@TestTransaction` interceptor binding, right? Ok, that's probably the simplest solution. However, I'd move as much as possible logic to the parent class, i.e. injection field of `UserTransaction` and the whole `@AroundInvoke` method. If it doesn't work then we have a bug in ArC...

stuartwdouglas replied:
It didn't work, although I did not have time to look into why.

mkouba replied:
Maybe it's because the parent class is not part of the index?

mkouba reviewed (changes requested):
I like the idea!

Added some comments...

famod commented:
WDYT about flushing the `EntityManager` before rolling back? We actually do so in a custom rollback interceptor (some colleagues were used to test rollbacks from Spring).
Otherwise you might not catch certain problems in your tests that occur on flush (e.g. constraint violations etc.).

stuartwdouglas commented:
Done

mkouba commented:
@stuartwdouglas FYI I'm going to rebase this PR, resolve conflicts and find out why the logic in the parent class did now work...

mkouba committed:
@TestTransaction - move logic to the TestTransactionInterceptor class

- add basic support for interceptor's inheritance
- other unsupported use cases are described here:
https://github.com/quarkusio/quarkus/issues/11942

mkouba reviewed (approved):
Looks good from my POV.

famod commented on extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/runtime/interceptor/TestTransactionInterceptor.java:
```diff
@@ -0,0 +1,46 @@
+
+    static final List<TestTransactionCallback> CALLBACKS;
+
+    static {
```
Thanks for having considered my `flush()`-suggestion. Very cool approach with this new `TestTransactionInterceptor`!

I am wondering whether this block could kick in cases where it is not really needed? Usually `ServiceLoader` calls are not super-cheap.
I used to apply https://en.wikipedia.org/wiki/Initialization-on-demand_holder_idiom but I don't know whether this would make any difference here.

mkouba replied:
Hm, the callbacks are only loaded once per test config...

mkouba replied:
That said, we could easily turn the list into a `io.quarkus.arc.impl.LazyValue<List<TestTransactionCallback>>` ...

stuartwdouglas replied:
The current code is fine. ServiceLoader is expensive in terms of 'don't do this every request so you are doing it tens of thousands of times per second', its fine for this.

famod replied:
Ok, thanks for your feedback.

famod commented:
I suppose this is not going to be included in 1.8.0.Final?
